contrib/ibm-mq/main.go

```diff
@@ -0,0 +1,178 @@
+		sink,
+		cloudevents.Builder{
+			Source:    "ibm:mq",
+			EventType: "message queue item",
+		},
+	)
+
+	log.Printf("%+v\n", cfg)
+
+	// create IBM MQ channel definition
+	channelDefinition := ibmmq.NewMQCD()
+	channelDefinition.ChannelName = cfg.ChannelName
+	channelDefinition.ConnectionName = cfg.ConnectionName
+
+	// init connection security params
+	connSecParams := ibmmq.NewMQCSP()
+	connSecParams.AuthenticationType = ibmmq.MQCSP_AUTH_USER_ID_AND_PWD
+	connSecParams.UserId = cfg.UserID
+	connSecParams.Password = cfg.Password
+
+	// setup MQ connection params
+	connOptions := ibmmq.NewMQCNO()
+	connOptions.Options = ibmmq.MQCNO_CLIENT_BINDING
+	// connOptions.ApplName = "Triggermesh eventing"
+	connOptions.ClientConn = channelDefinition
+	connOptions.SecurityParms = connSecParams
+
+	// And now we can try to connect. Wait a short time before disconnecting.
+	qMgrObject, err := ibmmq.Connx(cfg.QueueManager, connOptions)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Printf("Connection to %s succeeded.\n", cfg.QueueManager)
+	defer disconnect(qMgrObject)
+
+	// Create the Object Descriptor that allows us to give the queue name
+	mqod := ibmmq.NewMQOD()
+
+	// We have to say how we are going to use this queue. In this case, to GET
+	// messages. That is done in the openOptions parameter.
+	openOptions := ibmmq.MQOO_INPUT_SHARED
+
+	// Opening a QUEUE (rather than a Topic or other object type) and give the name
+	mqod.ObjectType = ibmmq.MQOT_Q
+	mqod.ObjectName = cfg.QueueName
+
+	qObject, err := qMgrObject.Open(mqod, openOptions)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Println("Opened queue", qObject.Name)
+	defer close(qObject)
+	defer wg.Wait()
+
+	for {
+		// The GET requires control structures, the Message Descriptor (MQMD)
+		// and Get Options (MQGMO). Create those with default values.
+		msgDescriptor := ibmmq.NewMQMD()
+		msgOptions := ibmmq.NewMQGMO()
+
+		// The default options are OK, but it's always
+		// a good idea to be explicit about transactional boundaries as
+		// not all platforms behave the same way.
+		msgOptions.Options = ibmmq.MQGMO_NO_SYNCPOINT
+
+		// Set options to wait for a maximum of 3 seconds for any new message to arrive
+		msgOptions.Options |= ibmmq.MQGMO_WAIT
+		msgOptions.WaitInterval = 3 * 1000 // The WaitInterval is in milliseconds
+
+		// Create a buffer for the message data. This one is large enough
+		// for the messages put by the amqsput sample.
+		buffer := make([]byte, 1024)
+
+		// Now we can try to get the message
+		_, err := qObject.Get(msgDescriptor, msgOptions, buffer)
+		if err != nil {
+			mqret := err.(*ibmmq.MQReturn)
+			if mqret != nil && mqret.MQRC == ibmmq.MQRC_NO_MSG_AVAILABLE {
+				continue
+			}
+			break
+		}
+
+		wg.Add(1)
+		go func(md *ibmmq.MQMD, messageData []byte) {
+			defer wg.Done()
+			data := string(bytes.Trim(messageData, "\u0000"))
+			msg := MQMessage{
+				Message:     md,
+				MessageData: data,
+			}
+			log.Printf("New message: %s", data)
```

Comment:
two things:
1 - is it safe from a security perspective to log the message contents itself? seems risky
2 - since there should be a ton of these running in parallel, I wonder if it would be helpful to be able to correlate all log messages from within a single thread? Meaning, prefix the logs with some kind of unique ID per thread. Then someone can correlate the error on line 156 with the incoming message. Of course, if you kill the above log then this isn't needed  :-)

Reply:
Removed message logging - it was there for debug purpose ;)